src/CloudNative.CloudEvents/Http/HttpClientExtensions.cs

```diff
@@ -286,6 +287,88 @@ public static bool IsCloudEventBatch(this HttpResponseMessage httpResponseMessag
+
+            byte[] content;
+            // The content type to include in the ContentType header - may be the data content type, or the formatter's content type.
+            ContentType contentType;
+            switch (contentMode)
+            {
+                case ContentMode.Structured:
+                    content = formatter.EncodeStructuredModeMessage(cloudEvent, out contentType);
+                    break;
+                case ContentMode.Binary:
+                    content = formatter.EncodeBinaryModeEventData(cloudEvent);
+                    contentType = MimeUtilities.CreateContentTypeOrNull(cloudEvent.DataContentType);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(contentMode), $"Unsupported content mode: {contentMode}");
+            }
+            var ret = new ByteArrayContent(content);
+            if (contentType is object)
+            {
+                ret.Headers.ContentType = MimeUtilities.ToMediaTypeHeaderValue(contentType);
+            }
+            else if (content.Length != 0)
+            {
+                throw new ArgumentException(Strings.ErrorContentTypeUnspecified, nameof(cloudEvent));
+            }
+
+            // Map headers in either mode.
+            // Including the headers in structured mode is optional in the spec (as they're already within the body) but
+            // can be useful.
+            ret.Headers.Add(HttpUtilities.SpecVersionHttpHeader, HttpUtilities.EncodeHeaderValue(cloudEvent.SpecVersion.VersionId));
+            foreach (var attributeAndValue in cloudEvent.GetPopulatedAttributes())
+            {
+                CloudEventAttribute attribute = attributeAndValue.Key;
+                string headerName = HttpUtilities.HttpHeaderPrefix + attribute.Name;
+                object value = attributeAndValue.Value;
+
+                // Skip the data content type attribute in binary mode, because it's already in the content type header.
+                if (attribute == cloudEvent.SpecVersion.DataContentTypeAttribute && contentMode == ContentMode.Binary)
+                {
+                    continue;
+                }
+                else
+                {
+                    string headerValue = HttpUtilities.EncodeHeaderValue(attribute.Format(value));
+                    ret.Headers.Add(headerName, headerValue);
+                }
+            }
+            return ret;
+        }
+
+        /// <summary>
+        /// Converts a CloudEvent batch to <see cref="HttpContent"/>.
+        /// </summary>
+        /// <param name="cloudEvents">The CloudEvent batch to convert. Must not be null, and every element must be non-null reference to a valid CloudEvent.</param>
+        /// <param name="formatter">The formatter to use within the conversion. Must not be null.</param>
+        public static HttpContent ToHttpContent(this IReadOnlyList<CloudEvent> cloudEvents, CloudEventFormatter formatter)
```

Comment:
The benefit of having a list is that we can validate all the events first before doing anything else - without risking (at least generally) generating two sets of events. Happy to talk about it further (there are definitely pros and cons) but I'll merge as-is for now.

Reply:
Agreed, I wouldn't risk enumerating them twice, but we could just do ToList immediately and work on that, though it would mean additional allocations.